op-e2e/actions/span_batch_test.go

```diff
@@ -114,6 +114,116 @@ func TestDropSpanBatchBeforeHardfork(gt *testing.T) {
+// If Delta activation time is in the middle of time range of a SpanBatch, op-node must drop the batch.
+func TestHardforkMiddleOfSpanBatch(gt *testing.T) {
+	t := NewDefaultTesting(gt)
+	p := &e2eutils.TestParams{
+		MaxSequencerDrift:   20, // larger than L1 block time we simulate in this test (12)
+		SequencerWindowSize: 24,
+		ChannelTimeout:      20,
+		L1BlockTime:         12,
+	}
+	dp := e2eutils.MakeDeployParams(t, p)
+
+	// Activate HF in the middle of the first epoch
+	deltaOffset := hexutil.Uint64(6)
+	dp.DeployConfig.L2GenesisDeltaTimeOffset = &deltaOffset
+	sd := e2eutils.Setup(t, dp, defaultAlloc)
+	log := testlog.Logger(t, log.LvlError)
+	miner, seqEngine, sequencer := setupSequencerTest(t, sd, log)
+	verifEngine, verifier := setupVerifier(t, sd, log, miner.L1Client(t, sd.RollupCfg), &sync.Config{})
+	minerCl := miner.EthClient()
+
+	rollupSeqCl := sequencer.RollupClient()
+	dp2 := e2eutils.MakeDeployParams(t, p)
+	minTs := hexutil.Uint64(0)
+	// Activate Delta hardfork for batcher. so batcher will submit SpanBatches to L1.
+	dp2.DeployConfig.L2GenesisDeltaTimeOffset = &minTs
+	sd2 := e2eutils.Setup(t, dp2, defaultAlloc)
+	batcher := NewL2Batcher(log, sd2.RollupCfg, &BatcherCfg{
+		MinL1TxSize: 0,
+		MaxL1TxSize: 128_000,
+		BatcherKey:  dp.Secrets.Batcher,
+	}, rollupSeqCl, miner.EthClient(), seqEngine.EthClient(), seqEngine.EngineClient(t, sd.RollupCfg))
+
+	// Alice makes a L2 tx
+	cl := seqEngine.EthClient()
+	n, err := cl.PendingNonceAt(t.Ctx(), dp.Addresses.Alice)
+	require.NoError(t, err)
+	signer := types.LatestSigner(sd.L2Cfg.Config)
+	tx := types.MustSignNewTx(dp.Secrets.Alice, signer, &types.DynamicFeeTx{
+		ChainID:   sd.L2Cfg.Config.ChainID,
+		Nonce:     n,
+		GasTipCap: big.NewInt(2 * params.GWei),
+		GasFeeCap: new(big.Int).Add(miner.l1Chain.CurrentBlock().BaseFee, big.NewInt(2*params.GWei)),
+		Gas:       params.TxGas,
+		To:        &dp.Addresses.Bob,
+		Value:     e2eutils.Ether(2),
+	})
+	require.NoError(gt, cl.SendTransaction(t.Ctx(), tx))
+
+	sequencer.ActL2PipelineFull(t)
+	verifier.ActL2PipelineFull(t)
+
+	miner.ActEmptyBlock(t)
+	sequencer.ActL1HeadSignal(t)
+
+	// Make a L2 block with the TX
+	sequencer.ActL2StartBlock(t)
+	seqEngine.ActL2IncludeTx(dp.Addresses.Alice)(t)
+	sequencer.ActL2EndBlock(t)
+
+	// HF is not activated yet
+	unsafeOriginNum := new(big.Int).SetUint64(sequencer.L2Unsafe().L1Origin.Number)
+	unsafeHeader, err := minerCl.HeaderByNumber(t.Ctx(), unsafeOriginNum)
+	require.NoError(t, err)
+	require.False(t, sd.RollupCfg.IsDelta(unsafeHeader.Time))
+
+	// Make L2 blocks until the next epoch
+	sequencer.ActBuildToL1Head(t)
+
+	// HF is activated for the last unsafe block
+	unsafeOriginNum = new(big.Int).SetUint64(sequencer.L2Unsafe().L1Origin.Number)
+	unsafeHeader, err = minerCl.HeaderByNumber(t.Ctx(), unsafeOriginNum)
+	require.NoError(t, err)
+	require.True(t, sd.RollupCfg.IsDelta(unsafeHeader.Time))
+
+	// Batch submit to L1. batcher should submit span batches.
+	batcher.ActSubmitAll(t)
```

Comment:
Why does it submit a span batch when it's not valid? Is that just a quirk of the test action implementation or would the real batcher actually get this wrong?

It feels a little weird to depend on the test code doing things wrong. I wonder if we should add something to explicitly force the batcher to use span batches similar to how it can submit garbage with `s.l2BatcherCfg.GarbageCfg`.